enhancements/authentication/improved-login-workflow.md

```diff
@@ -0,0 +1,348 @@
+and [PKCE](https://datatracker.ietf.org/doc/html/rfc7636).
+
+## Motivation
+
+Currently, when logging in with the `oc` client a user has to perform the following steps:
+
+1. Open the console in the browser.
+2. Login to console with the configured identity provider.
+3. Click on the **"Copy Login Command"** button which redirects the page to the OAuth application.
+4. Perform login again like in step 2.
+5. Copy the displayed login command.
+6. Run the copied command to configure the client to configure the client and access the cluster.
+
+This process can be simplified to the following steps:
+
+1. User runs `oc login --auto $CLUSTER_API_URL`
+2. `oc` starts a webserver to receive a callback from the OAuth server. It then opens the authorization endpoint with a
+   PKCE code challenge.
+3. The OAuth application performs the authentication either directly or by redirecting the user.
+4. Once authentication is successful the OAuth application redirects the user back to the callback URL of `oc` along
+   with a code.
+5. `oc` can then use this code along with the code challenge to request an access token. On receiving the token `oc`
+   then performs any configuration required to use the token.
+
+Using the above flow the user has to perform a login only once and the `oc` tool is configured automatically without any
+user input.
+
+### Goals
+
+Provide a one-step login for users which can be specified by the cluster operator.
+
+### Non-Goals
+
+TBD
+
+## Proposal
+
+### User Stories
+
+TBD
+
+### Implementation Details [optional]
+
+#### _oc login_ command
+
+The oc login command will have new 2 new flags, `--auto` and `--oauth-url`. The first flag indicates to `oc` that it
+should use the Oauth2 Authorization Flow to login. The second flag is optional and should be specified when the domain
+of the internal OAuth server has
+been [customized](https://docs.openshift.com/container-platform/4.9/authentication/configuring-internal-oauth.html#customizing-the-oauth-server-url_configuring-internal-oauth)
+. It will be the URL of the modified OAuth server.
+
+#### OAuth server
+
+The internal OAuth server uses [OSIN](https://github.com/openshift/osin) as a dependency to provide OAuth2 Authorized
+Grant Flow for a registered client to fetch tokens. It also has support for PKCE(Proof Key for Code Exchange) as
+specified in [rfc7636](https://datatracker.ietf.org/doc/html/rfc7636). _rfc7636_ is an enhancement for public clients,
+i.e. clients which cannot hold client credentials to fetch an access token securely. The new `oc login` is also a public
+client because it cannot contain client credentials to retrieve the access token and PKCE is designed for use cases like
+this one.
+
+Another change which may be required is to allow the port on the callback URL to be variable when the callback is to
+localhost. This is required because the webserver might not always be able to use the same port to listen on for the
+callback webserver.
+
```

Comment:
Do we need to consider token expiry ? -  I would imagine that the token is stored in memory within the "oc client" and not persisted.

Reply:
the token / code that comes back from the callback has a very short expiry usually, that's something we should clarify on how long.

As for the actual access token: doesn't it already have an expiration? This is controlled by the oauth server normally.

Reply:
the code currently has 5min expiry by default today, the actual access token expires in a day by default

Reply:
You can limit the life time of an access token by client type / trust:
- highest: client with credentials
- medium: native apps
- lowest: SPAs